quixstreams/app.py

```diff
@@ -75,6 +75,78 @@ def __call__(
     ) -> TopicManager: ...
 
 
+class RunTracker:
+    __slots__ = ("running", "_start_time", "_timeout", "_stop_checker", "_run_count")
+
+    _timeout_default = 0.0
+
+    def __init__(self):
+        """
+        Tracks the runtime status of an Application, along with managing variables
+        associated with stopping the app based on a timeout
+        """
+        self.running: bool = False
+        self._start_time: float = time.monotonic()
+        self._timeout: float = self._timeout_default
+        self._run_count: int = 0
+        self._stop_checker: Optional[Callable[[], bool]] = None
+
+    @property
+    def run_count(self) -> int:
+        """
+        Tracks number of times app.run() has been called.
+        Mostly only here for tracking the initial run.
+        :return:
+        """
+        return self._run_count
+
+    def update_status(self):
+        """
+        Trigger stop if its condition is met.
+        """
+        # most optimal way to keep performance with typical operation (no timeout)
+        if stopper := self._stop_checker:
+            if stopper():
+                self.stop()
+
+    def start(self):
+        """
+        Called as part of app.run() to initialize/reset stop conditions.
+        """
+        self.running = True
+        self._run_count += 1
+        self.reset_tracking()
+        if self._timeout:
+            self._stop_checker = self._at_timeout
+
+    def stop(self):
+        """
+        Called when Application is stopped, or self._stop_checker condition is met.
+        :return:
+        """
+        self.running = False
+        self._timeout = self._timeout_default
+
+    def set_stop_condition(self, timeout: Optional[float] = None):
+        """
+        Called as part of app.run(), where user can pass their time limit.
+        """
+        if timeout := (timeout or self._timeout_default):
```

Comment:
I think you can take a shortcut here (if you decide to change the default to 0.0 in the method signature) or is there something I'm not seeing?
```suggestion
        if self._timeout := timeout:
```

Reply:
Unfortunately, you can't set class attributes with a walrus operator, only variables.

In either case I changed the way this was setup so it doesn't need it now anyway =)